Use pathlib.Path for paths

keszybz opened the pull request.

behrmann reviewed (commented):
Haven't fully looked through the PR yet, only through (most) of the `Path` related changes (not the argument parser stuff). This is a very welcome change and looks good.

keszybz commented on mkosi/__init__.py:
```diff
@@ -1716,24 +1710,27 @@ def clean_yum_metadata(root: str) -> None:
-    keep_tdnf_data = os.access(tdnf_path, os.F_OK, follow_symlinks=False)
+    # fmt: off
+    tdnf_metadata_paths: List[PathString] = [
+        *glob.glob(f"{root / 'var/log/tdnf'}.*"),
```
Oh, indeed. I wasn't aware of `Path.glob`.

keszybz commented:
Rebased and updated with the two suggestions.

keszybz commented:
I noticed a typo and unnecessary type annotation in `clean_dnf/tdnf/rpm/yum_metadata()`. No other changes.

keszybz commented:
Another fixlet: str→Path conversion was missing in one place.

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3255,39 +3245,39 @@ def install_boot_loader(
+                copy_path(tree, root)
             else:
-                shutil.unpack_archive(d, root)
+                shutil.unpack_archive(cast(str, tree), root)
```
Why the cast here? Feels like this deserves a comment

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -3255,39 +3245,39 @@ def install_boot_loader(
+                copy_path(tree, root)
             else:
-                shutil.unpack_archive(d, root)
+                shutil.unpack_archive(str(tree), root)
```
Ditto

DaanDeMeyer commented on mkosi/__init__.py:
```diff
@@ -5143,29 +5175,15 @@ def parse_args(argv: Optional[List[str]] = None) -> Dict[str, argparse.Namespace
+        directory = args_pre_parsed.directory = args_pre_parsed.directory.absolute()
     else:
-        directory = os.path.abspath(".")
+        directory = Path(".").absolute()
```
You can simply put `Path()` here which defaults to the current directory

keszybz replied:
I actually changed that that Path.cwd(), it DTRT.

DaanDeMeyer commented on mkosi/backend.py:
```diff
@@ -395,50 +395,50 @@ def nspawn_params_for_blockdev_access(args: CommandLineArguments, loopdev: str)
     if result.returncode != 0:
         if "workspace-command" in ARG_DEBUG:
-            run(cmdline, check=False)
+            run(nspawn, check=False)  # XXX:???
```
I don't understand this comment at all

keszybz replied:
I was a note to myself, that I don't understand this part of the code. I dropped it now.

DaanDeMeyer reviewed (approved):
LGTM, although I'd appreciate if you could split out unrelated changes to separate commits next time. That would make reviewing a bit more manageable.

keszybz commented on mkosi/__init__.py:
```diff
@@ -529,13 +530,13 @@ def btrfs_subvol_delete(path: str) -> None:
             continue
         child_subvol_path = line.split(" ", 8)[-1]
-        child_path = os.path.normpath(os.path.join(path, cast(str, os.path.relpath(child_subvol_path, subvol_path))))
+        child_path = path / cast(str, os.path.relpath(child_subvol_path, subvol_path))
```
I used a cast because it's a noop: `os.path.relpath` returns `str`, but for some reason mypy doesn't understand that. `Path()` would actually create a new object, which isn't necessary. I expect that we'll just drop the cast later on.

keszybz commented on mkosi/__init__.py:
```diff
@@ -1381,26 +1381,26 @@ def install_etc_hostname(args: CommandLineArguments, root: str, cached: bool) ->
-        for d in paths:
-            mount_bind(d, root + d)
+        for subdir in subdirs:
+            mount_bind(Path(f"/{subdir}"), root / subdir)
```
Changed like that.

keszybz commented:
Updated. Apart from the changes discussed above, I changed the formatting of `add_argument()` to have all items one-per-line, for the items that were touched to add `type=Path`. (Before that'd require fmt:off fmt:on for black, but not it can just be done if desired.) This way those lines are much easier to read. No other changes.
